src_c/_sdl2/controller.pyx

```diff
@@ -0,0 +1,129 @@
+        not necessary to call this function.
+    """
+    SDL_GameControllerUpdate()
+
+def is_controller(index):
+    """ Check if the given joystick is supported by the game controller interface.
+
+    :param int index: Index of the joystick.
+
+    :return: 1 if supported, 0 if unsupported or invalid index.
+    """
+    return SDL_IsGameController(index)
+
+def name_forindex(index):
+    """ Returns the name of controller,
+        or NULL if there's no name or the index is invalid.
+    """
+    return SDL_GameControllerNameForIndex(index)
+
+cdef class Controller:
+    def __init__(self, int index):
+        """ Create a controller object and open it by given index.
+        
+        :param int index: Index of the joystick.
+        """
+
+        if not SDL_IsGameController(index):
+            raise error('Index is invalid or not a supported joystick.')
+
+        self._controller = SDL_GameControllerOpen(index)
+        self._index = index
+        if not self._controller:
+            raise error('Could not open controller %d.' % index)
+
+    @staticmethod
+    def from_joystick(joy):
+        # https://wiki.libsdl.org/SDL_GameControllerFromInstanceID
+        """ Create a controller object from pygame.joystick.Joystick object.
+        
+        """
+        if not pgJoystick_Check(joy):
+            raise TypeError('should be a pygame.joystick.Joystick object.')
+
+        cdef Controller self = Controller.__new__(Controller)
+        self.__init__(joy.get_id())
+        return self
+        
+    @staticmethod
+    def from_joystickid(SDL_JoystickID joyid):
+        """ Create a controller object from JoystickID
+        
+        """
+        # https://wiki.libsdl.org/SDL_GameControllerFromInstanceID
+        cdef Controller self = Controller.__new__(Controller)
+        self.__init__(joyid)
+        return self
+        
+    @property
+    def name(self):
+        # https://wiki.libsdl.org/SDL_GameControllerName
+        return SDL_GameControllerName(self._controller)
+    
+    def attached(self):
+        # https://wiki.libsdl.org/SDL_GameControllerGetAttached
+        return SDL_GameControllerGetAttached(self._controller)
+        
+    def as_joystick(self):
+        # return a pygame.joystick.Joystick() object.
+        joy = pgJoystick_New(self._index)
+        return joy
+        
+    def get_axis(self, SDL_GameControllerAxis axis):
+        # https://wiki.libsdl.org/SDL_GameControllerGetAxis
+        return SDL_GameControllerGetAxis(self._controller, axis)
+        
+    def get_button(self, SDL_GameControllerButton button):
+        # https://wiki.libsdl.org/SDL_GameControllerGetButton
+        return SDL_GameControllerGetButton(self._controller, button)
+        
+    def get_mapping(self):
+        #https://wiki.libsdl.org/SDL_GameControllerMapping
+        # TODO: mapping should be a readable dict instead of a string.
+        mapping = SDL_GameControllerMapping(self._controller)
+        SDL_free(mapping)
```

Comment:
This seems wrong. Either it's trying to free a Python string, or it may be using a string after it's freed. Maybe do this (untested):

```cython
mapping = SDL_GameControllerMapping(self._controller)
try:
    return mapping.decode('utf8')
finally:
    SDL_free(mapping)
```

Reply:
this causes unreachable code while compiling

I wonder if this works:
Convert it to a Python string and free the original mapping pointer.

        cdef char* mapping = SDL_GameControllerMapping(self._controller)
        
        try:
            #Convert the string to Python string?
            pyMapping = mapping.decode('UTF-8')
        except:
            #Returns bytes if decoding failed.
            pyMapping = mapping
        finally:
            SDL_free(mapping)
	        
        return pyMapping